### ISSUE TITLE
RFC: section records

### ISSUE THREAD

**Issue by andres-erbsen** [mit-plv/fiat-crypto#233]
I propose that we end every section with a record value where each field points to one definition in that section. Then another section importing this one can fill in the parameters for the record once in a vernac `Let` and use `record.(field)` syntax to access individual definitions.  The section record definition would always be a transparent application of the constructor for that record so uses of it would `iota`-reduce to pre-section-records code.

**Comment by JasonGross** [comment]
This only works nicely if the definitions are uniform in what section variables they access.  Importantly, if you mix proofs and definitions in the same section, they should go in different records.

**Comment by andres-erbsen** [comment]
Agreed. However, I would advocate for two sections in the case for which you suggested two records. In particular, the main reason some elliptic curve sections have proofs is to fill proj2_sig, so definitions need proof arguments anyway. Other than that, could we just have the proofs section reference the definitions section through its section record?

**Comment by andres-erbsen** [comment]
I think a downside of this approach is that definitions using a section with a section record will have two syntactically distinct ways of accessing definitions: through the section record and directly. I am considering whether it would be worth follow each section record specialization/import with ample `Let ... := Eval cbv iota in ... .`

**Comment by JasonGross** [comment]
Hmmm, maybe we should make all definitions in a section `Local Definition`s, and, when we define the record, run `Eval cbv beta delta [<list of all section definitions] in {| ... |}`.  This way, the non-section-record definitions will never be seen outside of the section they live in.  This is less important for section lemmas, which should never show up in the types of other things.

**Comment by andres-erbsen** [comment]
I am thinking I will actually go and try to convert the Curves directory to use section records instead of a multitude of bare Context variables.

What should we do about files that define both types and functions that use these types. The functions depend on the types up to conversion the same way as proofs depend on the functions up to conversion, this suggests that we actually may want *three* records: types, functions, proofs. In the case of ring/field, the types "record" would have only one type, in which case I think it may be fine to just use it without wrapping it in a record (but this convention would still force the carrier type of a ring to appear as a parameter of the record of ring operations, as opposed to as a field). This would allow users of these section records to require two different sets of operations on the same type... which may or may not make sense, but doesn't seem obviously wrong.

**Comment by andres-erbsen** [comment]
Here is an incarnation of how records like this might be defined to be used in lieu of module functors.

~~~~~~~~~~~~~~~~~~~~~~
Section WithEq.
  Context {T} (eq : T -> T -> Prop).
  Local Infix "=" := eq : type_scope.
  Class associative (op : T -> T -> T) :=
    { associativity : forall x y z, op x (op y z) = op (op x y) z }.
  Class commutative (op : T -> T -> T) :=
    { commutativity : forall x y, op x y = op y x }.
  Class is_left_identity_for (id:T) (op:T->T->T) :=
    { left_identity : forall x, op id x = x }.
  Class is_right_identity_for (id:T) (op:T->T->T) :=
    { right_identity : forall x, op x id = x }.
End WithEq.

Module Monoid.
  Class axioms {T} eq (id:T) op :=
    {
      associative : associative eq op;
      left_identity : is_left_identity_for eq id op;
      right_identity : is_right_identity_for eq id op;

      op_Proper: Proper (respectful eq (respectful eq eq)) op;
      Equivalence : Equivalence eq
    }.
  Global Existing Instance associative.
  Global Existing Instance left_identity.
  Global Existing Instance right_identity.
  Global Existing Instance Equivalence.
  Global Existing Instance op_Proper.

  Record definitions {T} :=
    {
      eq : T -> T -> Prop;
      id : T;
      op : T -> T -> T
    }.
  Arguments definitions : clear implicits.

  Class ok {T} (m:definitions T) :=
    { ok_axioms : axioms m.(eq) m.(id) m.(op) }.
  Global Existing Instance ok_axioms.
End Monoid.
~~~~~~~~~~~~~~~~~~~~~~

**Comment by andres-erbsen** [comment]
Motivation:

1. We need the version with all operations (`axioms`) as individual parameters to support code written before this set of definitions. For example, stdlib `Q` would satisfy `Field.axioms` but not `Field.ok` because its definitions are not projections from a `Field.definitions` record. All sections doing proofs about fields in general should use `axioms` and list out field operations individually instead of using `definitions`. Tactics should not assume a field is created using `definitions`.
2. We want `definitions` because we want to take it as a context variable in sections (e.g. elliptic curves parametrized over a field). Further, one section might take two different fields.
4. We want `T` to be a parameter of `definitions`, not a field, because we may want additional operations on the same type in some sections? Idk how common this is, but I am guessing it is necessary.
5. We want proofs to be separate from definitions, not only for build parallelism but also because a section may request multiple sets of proofs about the overlapping sets of definitions (e.g. a field where addition has order 3 or more).

**Comment by JasonGross** [comment]
Seems reasonable to me.  Things that come to mind:
- I think there's a paper on mathclasses that talks about unpacked vs packed records, and might be useful as a reference point
- Do we want to bind `=` to `@Monoid.eq _ _` in a particular scope?  If so, it might be useful to do something like
```coq
  Record definitions {T} :=
    {
      carrierT := T;
      eq : carrierT -> carrierT -> Prop;
      id : carrierT;
      op : carrierT -> carrierT -> carrierT
    }.
```
because then you have enough information to pick up what equality to use from an expression involving `op` and `id`.  (Then again, if we only have one monoid at a time, this may not be necessary.)

There's also a sort of diamond issue here: when stating a field, I not only want the same carrier type for the multiplicative group and the additive ring, I also want the same equality for both of them.  More generally, if we look to setoids for inspiration, the equality should always be bundled with the carrier type.  Alternatively, we're committing at the start to never nest `definition` records (nor, probably, `ok` records), and instead inline them and provide coercions or manual projection functions.  Is this what we want?

**Comment by andres-erbsen** [comment]
I agree that the mathcomp style makes a lot of sense if it is acceptable to require all code to be written with one particular algebraic hierarchy in mind; I am considering requiring code to have projections from a record to be off limits here. More discussion needed, it seems...

**Comment by JasonGross** [comment]
From discussion on signal: if we want modularity by transporting across isomorphisms, see also
- https://github.com/HoTT/HoTT/blob/master/theories/Tactics/EquivalenceInduction.v
- https://www.irit.fr/TYPES2013/Slides/TYPES13Slides_Denes_et_al.pdf

**Comment by andres-erbsen** [comment]
Here is why I wanted to put `T` as a record parameter but `eq : T -> T -> Prop` as a field: in a section parametrizing over two algebraic structures on `T`, it is possible to require `forall a b, eq1 a b <-> eq2 a b` by `T1 = T2` is not actually usable.

**Comment by andres-erbsen** [comment]
There is a bunch of good modularity/inheritance discussion at https://github.com/leanprover/lean/wiki/Refactoring-structures as well.
